Added last seen to profile view

canastillo opened the pull request:
For enhancement #2573 (closes #2573)

I'm having a problem with the timestamps. I'm getting dates from 1970. 

This is how it looks at the moment
![image](https://user-images.githubusercontent.com/66333350/170566901-56f1b28a-b887-4db5-a673-9aeadefe32bc.png)

Any suggestions?

Simon-Laux commented:
> Any suggestions?

Screenshot shows 1970 date, better hide the text or say "long time ago" / "unknown" if timestamp is equal to `0`

Also we could discuss the location, like android does it in a section bellow the name, similar to the signature, but I don't have a strong opinion on that, maybe its fine as is (directly under the name)

adbenitez commented:
> I'm having a problem with the timestamps. I'm getting dates from 1970.

hola Ana, gracias por contribuir,

dates from 1970 mean that you have never seen that contact in that case `Last seen: Unknown` must be shown (`last_seen_unknown` string)

adbenitez commented:
oops! @Simon-Laux  already replied

canastillo commented:
Thanks for your comments. I have made some changes for that case with the 70s. I wrote the condition as `<= 1970` because I remember seing even a date from 1969. 

I just have two questions: 
- Should I consider a later year or is 1970 okay?
- When does this "unknown" case happens? Because I chatted with the person of the screenshot like two days ago and I chatted today with another one and the same thing happens. Last seen has dates from 1970.

About the location, I'm open to suggestions. I thought that spot was the right place but I can change it. No problem.

Jikstra commented:
> Thanks for your comments. I have made some changes for that case with the 70s. I wrote the condition as `<= 1970` because I remember seing even a date from 1969 (Dec 31st probably). I didn't write anything about `lastSeen == 0` 'cause I'm actually not getting 0s from the backend, but small numbers like 1653452517 or 1653437084.
> 
> I just have two questions:
> 
> * Should I consider a later year or is 1970 okay?
> * When does this "unknown" case happens? Because I chatted with the person of the screenshot like two days ago and I chatted today with another one and the same thing happens. Last seen has dates from 1970.
> 
> About the location, I'm open to suggestions. I thought that spot was the right place but I can change it. No problem.

I have the feeling it's something else, @adbenitez in theory 0 should be returned if there's no seen info right?

Jikstra commented on src/renderer/components/dialogs/ViewProfile.tsx:
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
         <p className='group-name'>{name}</p>
       </div>
       <div className='address'>{address}</div>
+      <div>
+        <p>{`${lastSeenString}`}</p>
+      </div>
```
```suggestion
      <div className='last-seen'>{lastSeenString}`}</div>
```

Jikstra commented on src/renderer/components/dialogs/ViewProfile.tsx:
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
+  else {
+    const extended = false
+    const date = formatRelativeTime(lastSeen, { extended })
+    lastSeenString = tx('last_seen_at', `${date}`)
```
```suggestion
    lastSeenString = tx('last_seen_at', date)
```

Jikstra commented on src/renderer/components/dialogs/ViewProfile.tsx:
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
+
+  // Dates from 1970 mean that contact has never been seen
+  if (parseInt(year) <= 1970) lastSeenString = tx('last_seen_unknown')
+  else {
```
```suggestion
  if (parseInt(year) <= 1970) {
    lastSeenString = tx('last_seen_unknown')
  } else {
```

Jikstra commented on src/renderer/components/dialogs/ViewProfile.tsx:
```diff
@@ -30,16 +32,33 @@ import { selectChat } from '../helpers/ChatMethods'
+  let lastSeenString = ''
+
+  // Dates from 1970 mean that contact has never been seen
+  if (parseInt(year) <= 1970) lastSeenString = tx('last_seen_unknown')
```
```
  if (parseInt(year) <= 1970) {
    lastSeenString = tx('last_seen_unknown')
  } else {
```

canastillo replied:
Can you please write this as a suggestion so I can add it to the batch? The button doesn't appear in this one

Jikstra reviewed (commented):
Added some small code changes :)

Simon-Laux commented:
It turns out the timestamp from the core is in seconds, while the timestamp expected by momenjs is in milliseconds, so you need to do `lastSeen * 1000`:
const date = formatRelativeTime(lastSeen * 1000, { extended })

then you can also just compare to `0` instead of parsing the year. we also compare timestamps to `0` in other places, recently I added it for the chatlistitems of chats that have no messages yet.

After fixing it locally I noticed this:
<img width="142" alt="image" src="https://user-images.githubusercontent.com/18725968/170692920-7532e519-95be-4d9b-82a3-a5939d30dd4e.png">

which makes not much sense to me, so I think we should use another relative time format, so that it says "Monday" instead of just "Mon". Or we opt for just the absolute date/time? (has is it done on android @adbenitez ?)

adbenitez commented:
> I have the feeling it's something else, @adbenitez in theory 0 should be returned if there's no seen info right?

yes, and that is what android checks for: 
https://github.com/deltachat/deltachat-android/blob/master/src/org/thoughtcrime/securesms/ProfileSettingsAdapter.java#L240

adbenitez commented:
> which makes not much sense to me, so I think we should use another relative time format, so that it says "Monday" instead of just "Mon". Or we opt for just the absolute date/time? (has is it done on android @adbenitez ?)

for how it is done in android check:
https://github.com/deltachat/deltachat-android/blob/b488103c43fdd5c7dfa059cc1c6e33c1f0b813dd/src/org/thoughtcrime/securesms/util/DateUtils.java#L80

```
  public static String getExtendedTimeSpanString(final Context c, final Locale locale, final long timestamp) {
    StringBuilder format = new StringBuilder();
    if      (DateUtils.isToday(timestamp))                 {}
    else if (isWithin(timestamp,   6, TimeUnit.DAYS)) format.append("EEE ");
    else if (isWithin(timestamp, 365, TimeUnit.DAYS)) format.append("MMM d, ");
    else                                              format.append("MMM d, yyyy, ");

    if (DateFormat.is24HourFormat(c)) format.append("HH:mm");
    else                              format.append("hh:mm a");

    return getFormattedDateTime(timestamp, format.toString(), locale);
  }
```

Jikstra commented on src/renderer/components/dialogs/ViewProfile.tsx:
```diff
@@ -30,16 +32,31 @@ import { selectChat } from '../helpers/ChatMethods'
+  if (lastSeen == 0) {
+    lastSeenString = tx('last_seen_unknown')
+  } else {
+    const extended = false
+    const date = formatRelativeTime(lastSeen * 1000, { extended })
```
```suggestion
    const date = formatRelativeTime(lastSeen * 1000, { extended: false })
```

Jikstra replied:
One last small code nitpick and then I'm happy to merge :P

canastillo replied:
Done! :D

canastillo commented:
Thanks @Simon-Laux for the explanation. I now understand what is going on. I changed that conditional and added the changes that Jikstra made. 

So, do you want me to add another relative format? I think on android they use the format that is used already in here.

Jikstra commented:
> So, do you want me to add another relative format?

I wouldn't do it in this pr, but do we need another relative format? What was the discussion on this? Probably I missed it :P
